apimachinery + client-go + device taint eviction unit test: context-aware Start/WaitFor, waiting through channels

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Context-aware SharedInformerFactory.Start is best explained by looking at the output of the `pkg/controller/devicetainteviction` unit test when a single unit test is intentionally broken - yes, it's long, I dare you to look at the attached file...
[before.txt](https://github.com/user-attachments/files/23681951/before.txt)

With this PR, it becomes much shorter because global log spam on stderr is avoided through (more or less) consistent usage of per-test output:

```
I1121 18:50:28.112284  396950 feature_gate.go:466] feature gates: {map[DRADeviceTaintRules:true DRADeviceTaints:true]}
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:10:00.020000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.010000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:30.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:30.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:30.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:30.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:30.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I0101 01:00:00.000000  396950 watch.go:218] "Stopping fake watcher"
I1121 18:50:29.704907  396950 feature_gate.go:466] feature gates: {map[DRADeviceTaintRules:false DRADeviceTaints:false]}
--- FAIL: TestAll (1.58s)
    --- FAIL: TestAll/Eviction (0.02s)
        --- FAIL: TestAll/Eviction/initial (0.00s)
            device_taint_eviction_test.go:2132: I0101 01:00:00.000000] starting to wait for watches
            reflector.go:314: I0101 01:00:00.000000] k8s.io/client-go/informers/factory.go:163: The provided ListWatcher doesn't support WatchList semantics. The feature will be disabled. If you are using a custom client, check the documentation of watchlist.DoesClientNotSupportWatchListSemantics() method listWatcherType="*cache.listWatcherWithWatchListSemanticsWrapper" feature=<features.Feature>: WatchListClient
            reflector.go:314: I0101 01:00:00.000000] k8s.io/client-go/informers/factory.go:163: The provided ListWatcher doesn't support WatchList semantics. The feature will be disabled. If you are using a custom client, check the documentation of watchlist.DoesClientNotSupportWatchListSemantics() method listWatcherType="*cache.listWatcherWithWatchListSemanticsWrapper" feature=<features.Feature>: WatchListClient
            reflector.go:314: I0101 01:00:00.000000] k8s.io/client-go/informers/factory.go:163: The provided ListWatcher doesn't support WatchList semantics. The feature will be disabled. If you are using a custom client, check the documentation of watchlist.DoesClientNotSupportWatchListSemantics() method listWatcherType="*cache.listWatcherWithWatchListSemanticsWrapper" feature=<features.Feature>: WatchListClient
            reflector.go:377: I0101 01:00:00.000000] Starting reflector type="*v1.ResourceSlice" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:421: I0101 01:00:00.000000] Listing and watching type="*v1.ResourceSlice" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:453: I0101 01:00:00.000000] Caches populated type="*v1.ResourceSlice" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:377: I0101 01:00:00.000000] Starting reflector type="*v1.DeviceClass" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:377: I0101 01:00:00.000000] Starting reflector type="*v1alpha3.DeviceTaintRule" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:421: I0101 01:00:00.000000] Listing and watching type="*v1alpha3.DeviceTaintRule" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:453: I0101 01:00:00.000000] Caches populated type="*v1alpha3.DeviceTaintRule" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:314: I0101 01:00:00.000000] k8s.io/client-go/informers/factory.go:163: The provided ListWatcher doesn't support WatchList semantics. The feature will be disabled. If you are using a custom client, check the documentation of watchlist.DoesClientNotSupportWatchListSemantics() method listWatcherType="*cache.listWatcherWithWatchListSemanticsWrapper" feature=<features.Feature>: WatchListClient
            reflector.go:377: I0101 01:00:00.000000] Starting reflector type="*v1.ResourceClaim" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:421: I0101 01:00:00.000000] Listing and watching type="*v1.ResourceClaim" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:421: I0101 01:00:00.000000] Listing and watching type="*v1.DeviceClass" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:453: I0101 01:00:00.000000] Caches populated type="*v1.ResourceClaim" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:453: I0101 01:00:00.000000] Caches populated type="*v1.DeviceClass" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:314: I0101 01:00:00.000000] k8s.io/client-go/informers/factory.go:163: The provided ListWatcher doesn't support WatchList semantics. The feature will be disabled. If you are using a custom client, check the documentation of watchlist.DoesClientNotSupportWatchListSemantics() method listWatcherType="*cache.listWatcherWithWatchListSemanticsWrapper" feature=<features.Feature>: WatchListClient
            reflector.go:377: I0101 01:00:00.000000] Starting reflector type="*v1.Pod" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:421: I0101 01:00:00.000000] Listing and watching type="*v1.Pod" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:453: I0101 01:00:00.000000] Caches populated type="*v1.Pod" reflector="k8s.io/client-go/informers/factory.go:163"
            device_taint_eviction_test.go:2135: I0101 01:00:00.000000] done waiting for watches
            device_taint_eviction.go:761: I0101 01:00:00.000000] Starting controller="device-taint-eviction" numWorkers=10
            device_taint_eviction.go:801: I0101 01:00:00.000000] Sending events to api server
            shared_informer.go:438: I0101 01:00:00.000000] Waiting for="cache and event handler sync"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.Pod"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.ResourceClaim"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.ResourceSlice"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.DeviceClass"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1alpha3.DeviceTaintRule"
            device_taint_eviction.go:1089: I0101 01:00:00.000000] ResourceClaim changed claimObject="default/my-pod-my-resource" oldClaim=<
                	null
                 > newClaim=<
                	{
                	  "metadata": {
                	    "name": "my-pod-my-resource",
                	    "namespace": "default",
                	    "ownerReferences": [
                	      {
                	        "apiVersion": "v1",
                	        "kind": "Pod",
                	        "name": "my-pod",
                	        "uid": "1234",
                	        "controller": true
                	      }
                	    ],
                	    "finalizers": [
                	      "resource.kubernetes.io/delete-protection"
                	    ]
                	  },
                	  "spec": {
                	    "devices": {
                	      "requests": [
                	        {
                	          "name": "req-1",
                	          "exactly": {
                	            "deviceClassName": "my-resource-class",
                	            "allocationMode": "ExactCount",
                	            "count": 1
                	          }
                	        }
                	      ]
                	    }
                	  },
                	  "status": {
                	    "allocation": {
                	      "devices": {
                	        "results": [
                	          {
                	            "request": "req-1",
                	            "driver": "some-driver",
                	            "pool": "worker",
                	            "device": "instance"
                	          }
                	        ]
                	      }
                	    },
                	    "reservedFor": [
                	      {
                	        "resource": "pods",
                	        "name": "my-pod",
                	        "uid": "1234"
                	      }
                	    ]
                	  }
                	}
                 > diff=<
                	@@ -1 +1,53 @@
                	-null
                	+{
                	+ "metadata": {
                	+  "name": "my-pod-my-resource",
                	+  "namespace": "default",
                	+  "ownerReferences": [
                	+   {
                	+    "apiVersion": "v1",
                	+    "kind": "Pod",
                	+    "name": "my-pod",
                	+    "uid": "1234",
                	+    "controller": true
                	+   }
                	+  ],
                	+  "finalizers": [
                	+   "resource.kubernetes.io/delete-protection"
                	+  ]
                	+ },
                	+ "spec": {
                	+  "devices": {
                	+   "requests": [
                	+    {
                	+     "name": "req-1",
                	+     "exactly": {
                	+      "deviceClassName": "my-resource-class",
                	+      "allocationMode": "ExactCount",
                	+      "count": 1
                	+     }
                	+    }
                	+   ]
                	+  }
                	+ },
                	+ "status": {
                	+  "allocation": {
                	+   "devices": {
                	+    "results": [
                	+     {
                	+      "request": "req-1",
                	+      "driver": "some-driver",
                	+      "pool": "worker",
                	+      "device": "instance"
                	+     }
                	+    ]
                	+   }
                	+  },
                	+  "reservedFor": [
                	+   {
                	+    "resource": "pods",
                	+    "name": "my-pod",
                	+    "uid": "1234"
                	+   }
                	+  ]
                	+ }
                	+}
                 >
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.ResourceClaim + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.Pod + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1alpha3.DeviceTaintRule + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"
            device_taint_eviction.go:1328: I0101 01:00:00.000000] ResourceSlice changed pool=<devicetainteviction.poolID>: {
                        driverName: "some-driver",
                        poolName: "worker",
                    } oldSlice=<
                	null
                 > newSlice=<
                	{
                	  "metadata": {
                	    "name": "worker-some-driver"
                	  },
                	  "spec": {
                	    "driver": "some-driver",
                	    "pool": {
                	      "name": "worker",
                	      "generation": 0,
                	      "resourceSliceCount": 1
                	    },
                	    "nodeName": "worker",
                	    "devices": [
                	      {
                	        "name": "instance",
                	        "taints": [
                	          {
                	            "key": "example.com/taint",
                	            "value": "something",
                	            "effect": "NoExecute",
                	            "timeAdded": "2000-01-01T00:00:00Z"
                	          }
                	        ]
                	      },
                	      {
                	        "name": "instance-no-schedule",
                	        "taints": [
                	          {
                	            "key": "example.com/taint",
                	            "effect": "NoSchedule"
                	          }
                	        ]
                	      },
                	      {
                	        "name": "instance-no-execute",
                	        "taints": [
                	          {
                	            "key": "example.com/taint",
                	            "effect": "None",
                	            "timeAdded": "2000-01-01T00:00:00Z"
                	          },
                	          {
                	            "key": "example.com/taint",
                	            "effect": "NoExecute",
                	            "timeAdded": "2000-01-01T00:00:00Z"
                	          }
                	        ]
                	      }
                	    ]
                	  }
                	}
                 > diff=<
                	@@ -1 +1,51 @@
                	-null
                	+{
                	+ "metadata": {
                	+  "name": "worker-some-driver"
                	+ },
                	+ "spec": {
                	+  "driver": "some-driver",
                	+  "pool": {
                	+   "name": "worker",
                	+   "generation": 0,
                	+   "resourceSliceCount": 1
                	+  },
                	+  "nodeName": "worker",
                	+  "devices": [
                	+   {
                	+    "name": "instance",
                	+    "taints": [
                	+     {
                	+      "key": "example.com/taint",
                	+      "value": "something",
                	+      "effect": "NoExecute",
                	+      "timeAdded": "2000-01-01T00:00:00Z"
                	+     }
                	+    ]
                	+   },
                	+   {
                	+    "name": "instance-no-schedule",
                	+    "taints": [
                	+     {
                	+      "key": "example.com/taint",
                	+      "effect": "NoSchedule"
                	+     }
                	+    ]
                	+   },
                	+   {
                	+    "name": "instance-no-execute",
                	+    "taints": [
                	+     {
                	+      "key": "example.com/taint",
                	+      "effect": "None",
                	+      "timeAdded": "2000-01-01T00:00:00Z"
                	+     },
                	+     {
                	+      "key": "example.com/taint",
                	+      "effect": "NoExecute",
                	+      "timeAdded": "2000-01-01T00:00:00Z"
                	+     }
                	+    ]
                	+   }
                	+  ]
                	+ }
                	+}
                 >
            device_taint_eviction.go:1205: I0101 01:00:00.000000] Claim is affected by device taint claim="default/my-pod-my-resource" device=<v1.DeviceRequestAllocationResult>: 
                        device: instance
                        driver: some-driver
                        pool: worker
                        request: req-1 taint="ResourceSlice /worker-some-driver some-driver/worker/instance taint #0" evictionTime="2000-01-01 00:00:00 +0000 UTC"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.ResourceSlice + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"
            device_taint_eviction.go:1001: I0101 01:00:00.000000] Underlying informers have synced
            device_taint_eviction_test.go:2333: I0101 01:00:00.000000] ERROR: 
                	Error Trace:	/nvme/gopath/src/k8s.io/kubernetes/pkg/controller/devicetainteviction/device_taint_eviction_test.go:2333
                	            				/nvme/gopath/src/k8s.io/kubernetes/test/utils/ktesting/tcontext.go:416
                	Error:      	Not equal: 
                	            	expected: 1
                	            	actual  : 0
                	Test:       	TestAll/Eviction/initial
                	Messages:   	get pod once
            device_taint_eviction_test.go:2334: I0101 01:00:00.000000] ERROR: 
                	Error Trace:	/nvme/gopath/src/k8s.io/kubernetes/pkg/controller/devicetainteviction/device_taint_eviction_test.go:2334
                	            				/nvme/gopath/src/k8s.io/kubernetes/test/utils/ktesting/tcontext.go:416
                	Error:      	Not equal: 
                	            	expected: 1
                	            	actual  : 0
                	Test:       	TestAll/Eviction/initial
                	Messages:   	update pod once
            device_taint_eviction_test.go:2335: I0101 01:00:00.000000] ERROR: 
                	Error Trace:	/nvme/gopath/src/k8s.io/kubernetes/pkg/controller/devicetainteviction/device_taint_eviction_test.go:2335
                	            				/nvme/gopath/src/k8s.io/kubernetes/test/utils/ktesting/tcontext.go:416
                	Error:      	Not equal: 
                	            	expected: 1
                	            	actual  : 0
                	Test:       	TestAll/Eviction/initial
                	Messages:   	delete pod once
            device_taint_eviction_test.go:2355: I0101 01:00:00.000000] ERROR: unexpected modified pod (-want, +got):
                  (*v1.Pod)(
                - 	s"&Pod{ObjectMeta:{my-pod  default  1234  0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:PodSpec{Volumes:[]Volume{},Containers:[]Container{},RestartPolicy:,TerminationGracePeriodSeconds:nil,ActiveDeadlineSeconds:nil,DNSPolicy:,NodeSel"...,
                + 	nil,
                  )
            device_taint_eviction_test.go:2366: I0101 01:00:00.000000] FATAL ERROR: 
                Expected
                    <[]v1.Event | len:0, cap:0>: nil
                to consist of
                    <[]*gstruct.FieldsMatcher | len:1, cap:1>: [
                        {
                            Fields: {
                                "Source": <*matchers.HaveFieldMatcher | 0xc008f977e0>{
                                    Field: "Component",
                                    Expected: <*matchers.EqualMatcher | 0xc001a9ea80>{
                                        Expected: <string>"device-taint-eviction",
                                    },
                                },
                                "Reason": <*matchers.EqualMatcher | 0xc001a9ea90>{
                                    Expected: <string>"DeviceTaintManagerEviction",
                                },
                                "Message": <*matchers.EqualMatcher | 0xc001a9eab0>{
                                    Expected: <string>"Marking for deletion",
                                },
                                "Count": <*matchers.EqualMatcher | 0xc001a9eac0>{Expected: <int32>1},
                                "InvolvedObject": <*matchers.EqualMatcher | 0xc001a9ead0>{
                                    Expected: 
                                        apiVersion: v1
                                        kind: Pod
                                        name: my-pod
                                        namespace: default
                                        uid: "1234",
                                },
                            },
                            IgnoreExtras: true,
                            IgnoreUnexportedExtras: false,
                            IgnoreMissing: false,
                            failures: nil,
                        },
                    ]
                the missing elements were
                    <[]*gstruct.FieldsMatcher | len:1, cap:1>: [
                        {
                            Fields: {
                                "Source": <*matchers.HaveFieldMatcher | 0xc008f977e0>{
                                    Field: "Component",
                                    Expected: <*matchers.EqualMatcher | 0xc001a9ea80>{
                                        Expected: <string>"device-taint-eviction",
                                    },
                                },
                                "Reason": <*matchers.EqualMatcher | 0xc001a9ea90>{
                                    Expected: <string>"DeviceTaintManagerEviction",
                                },
                                "Message": <*matchers.EqualMatcher | 0xc001a9eab0>{
                                    Expected: <string>"Marking for deletion",
                                },
                                "Count": <*matchers.EqualMatcher | 0xc001a9eac0>{Expected: <int32>1},
                                "InvolvedObject": <*matchers.EqualMatcher | 0xc001a9ead0>{
                                    Expected: 
                                        apiVersion: v1
                                        kind: Pod
                                        name: my-pod
                                        namespace: default
                                        uid: "1234",
                                },
                            },
                            IgnoreExtras: true,
                            IgnoreUnexportedExtras: false,
                            IgnoreMissing: false,
                            failures: nil,
                        },
                    ]
            device_taint_eviction_test.go:2312: I0101 01:00:00.000000] Waiting for goroutine termination...
            device_taint_eviction.go:794: I0101 01:00:00.000000] Shutting down work queue
            device_taint_eviction.go:1012: I0101 01:00:00.000000] Shut down controller controller="device-taint-eviction" reason=<nil>: nil
            reflector.go:383: I0101 01:00:00.000000] Stopping reflector type="*v1.DeviceClass" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:383: I0101 01:00:00.000000] Stopping reflector type="*v1.ResourceClaim" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:383: I0101 01:00:00.000000] Stopping reflector type="*v1.ResourceSlice" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:383: I0101 01:00:00.000000] Stopping reflector type="*v1.Pod" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
            reflector.go:383: I0101 01:00:00.000000] Stopping reflector type="*v1alpha3.DeviceTaintRule" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:163"
FAIL
FAIL	k8s.io/kubernetes/pkg/controller/devicetainteviction	1.621s
FAIL
```

Remaining culprits are `feature_gate.go` (only two lines now because the test was rewritten to set feature gates only once, but elsewhere the problem remains) and `watch.go`.

The new `cache.WaitFor` is able to describe what it is waiting for:

```
            shared_informer.go:438: I0101 01:00:00.000000] Waiting for="cache and event handler sync"
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.Pod"
...
            shared_informer.go:456: I0101 01:00:00.000000] Done waiting for="cache and event handler sync" instance="RealFIFO *v1.ResourceSlice + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"
```

Waiting for channels to close in `WaitFor` creates a causal relationship between goroutines which is visible to synctest. When a controller in a synctest bubble does a WaitFor in a test's background goroutine for the controller, the test can use `synctest.Wait` to wait for completion of cache sync, without requiring any test specific "has controller synced" API. Without this, the test had to poll or otherwise wait for the controller.

The polling in `WaitForCacheSync` moved the virtual clock forward by a random amount, depending on how often it had to check in wait.Poll. Now tests can be written such that all events during a test happen at a predictable time. This gets demonstrated in the updated pkg/controller/devicetainteviction unit test, addressing several open TODOs there (see https://github.com/kubernetes/kubernetes/pull/136939).

#### Which issue(s) this PR is related to:

Fixes: https://github.com/kubernetes/kubernetes/issues/135315

#### Special notes for your reviewer:

Review commit-by-commit might work better.

#### Does this PR introduce a user-facing change?

I'm intentionally not documenting the Go API changes here. Some changes are backward-compatible, but others like changing the AddEventHandler return type are not. In-tree some types emulated informers and had to be updated. This isn't a very common pattern, so the expectation is that most out-of-free users of informers do not need to be changed.

```release-note
NONE
```

/cc @jpbetz

pohly committed:
client-go cache: wait for cache sync via channels, better logging

The main advantage is that waiting on channels creates a causal relationship
between goroutines which is visible to synctest. When a controller in a
synctest bubble does a WaitFor in a test's background goroutine for the
controller, the test can use synctest.Wait to wait for completion of cache
sync, without requiring any test specific "has controller synced" API. Without
this, the test had to poll or otherwise wait for the controller.

The polling in WaitForCacheSync moved the virtual clock forward by a random
amount, depending on how often it had to check in wait.Poll. Now tests can be
written such that all events during a test happen at a predictable time. This
will be demonstrated in a separate commit for the
pkg/controller/devicetainteviction unit test.

The benefit for normal production is immediate continuation when the last
informer is synced (not really a problem, but still...) and more important,
nicer logging thanks to the names associated with the thing that is being
waited for. The caller decides whether logging is enabled or disabled and
describes what is being waited for (typically informer caches, but maybe also
event handlers or even something else entirely as long as it implements the
DoneChecker interface).

Before:

    Waiting for caches to sync
    Caches are synced

After:

    Waiting for="cache and event handler sync"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1.Pod"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1.ResourceClaim"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1.ResourceSlice"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1.DeviceClass"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1alpha3.DeviceTaintRule"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1.ResourceClaim + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1.Pod + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1alpha3.DeviceTaintRule + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"
    Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1.ResourceSlice + event handler k8s.io/kubernetes/pkg/controller/devicetainteviction.(*Controller).Run"

The "SharedIndexInformer *v1.Pod" is also how this appears in metrics.

pohly committed:
client-go cache: allow passing name+logger to DeltaFIFO, RealFIFO and Reflector

This improves logging and enables more informative waiting for cache sync in a
following commit. It addresses one klog.TODO in the Reflector.

The RealFIFOOptions and InformerOptions structs get extended the same way as
DeltaFIFOOptions before: a logger may be set, but it's not required. This is
not an API break.

That the name has to be passed separately is a bit annoying at first glance
because it could also be set directly on the logger through WithName, but
keeping it separate is better:
- name can be set without providing a logger
- name can be defaulted
- less code in the caller when passing through a logger and adding
  the name only in the field
- last but not least, extracting the name is not supported in a portable
  manner by logr

All in-tree references in production code get updated.

While at it, logging in the fifos gets updated to follow best practices: if
some code encounters an abnormal situation and then continues, it should use
utilruntime.HandleErrorWithLogger instead of normal error logging.

Existing "logger" fields get moved to the top because that is a more common
place for such a read-only field.

pohly committed:
client-go informers: context-aware Start + WaitForCacheSync

Passing a context to StartWithContext enables context-aware reflector
logging. This is the main remaining source of log spam (output to stderr
instead of per-test logger) in controller unit tests.

WaitForCacheSynceWithContext takes advantage of the new cache.WaitFor +
NamedHasSynced functionality to finish "immediately" (= no virtual time
passed) in a synctest bubble. While at it, the return type gets improved so
that a failure is easier to handle.

pohly commented on staging/src/k8s.io/client-go/informers/factory.go:
```diff
@@ -195,6 +197,81 @@ func (f *sharedInformerFactory) WaitForCacheSync(stopCh <-chan struct{}) map[ref
+}
+
+// CacheSyncResult is the result of WaitForCacheSyncWithContext.
+type CacheSyncResult struct {
```
I probably should move this to an apimachinery package. Then I can also add a unit test (missing right now).

jpbetz replied:
SG.  I was also noticing the duplicate implementations of `WaitForCacheSyncWithContext` here and in apiextensions.  Would be good to do something about that at some point.  I'm fine with a tracking issue for now.

pohly reviewed (commented):
Not completely finished, but I'd like to get feedback before I continue...

pohly commented:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/135395/pull-kubernetes-verify/1991933383281741824:
```
Test files may not access mutable global feature gates directly:
pkg/controller/devicetainteviction/device_taint_eviction_test.go:// we cannot change them in an init function with utilfeature.DefaultMutableFeatureGate.SetFromMap

Use this invocation instead:
  featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, <value>)
```

The first rule of feature gate testing: don't talk about `utilfeature.DefaultMutableFeatureGate`, not even in a comment...

pohly commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -399,6 +421,72 @@ func WaitForCacheSync(stopCh <-chan struct{}, cacheSyncs ...InformerSynced) bool
+// need polling, which makes it react immediately. When used in a synctest unit
+// test, waiting completes without moving randomly moving time forward, which
+// makes tests more predictable.
+func WaitFor(ctx context.Context, what string, cacheSyncs ...NamedSyncer) bool {
```
/hold

jpbetz replied:
FWIW, I favor the idea of putting the bulk of this in the `wait` package with appropriate aliases here.

michaelasp replied:
`WaitFor` is good, `NamedSyncer` feels a bit clunky and reads more like a worker than a way to read the state of the controller. If we ever extend it later it may not be a great name anymore. It also doesn't really explain the async portion of it, maybe `SyncHandle`/`NamedSyncHandle` to pass the idea along that it's a handle for figuring out what the current state of the informer is. Not too picky on this though, as long as it reads ok.

jpbetz reviewed (commented):
I didn't complete a full review but got my head around the approach and really like it in general.  Only one major concern from me so far.

pohly commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -141,6 +142,9 @@ type Controller interface {
 	HasSynced() bool
 
+	// NamedHasSynced delegates to the Config's Queue
+	NamedHasSynced() NamedSyncer
```
The interface doesn't seem to be used much even internally:

```
type SharedInformer interface {
...
	// GetController is deprecated, it does nothing useful
	GetController() Controller
```

Avoiding the API break turned out be trivial:
```patch
$ git diff
diff --git a/staging/src/k8s.io/client-go/tools/cache/controller.go b/staging/src/k8s.io/client-go/tools/cache/controller.go
index 0739e2c95e4..0fd595be289 100644
--- a/staging/src/k8s.io/client-go/tools/cache/controller.go
+++ b/staging/src/k8s.io/client-go/tools/cache/controller.go
@@ -119,6 +119,8 @@ type controller struct {
        clock          clock.Clock
 }
 
+var _ NamedController = &controller{}
+
 // Controller is a low-level controller that is parameterized by a
 // Config and used in sharedIndexInformer.
 type Controller interface {
@@ -142,16 +144,26 @@ type Controller interface {
        // HasSynced delegates to the Config's Queue
        HasSynced() bool
 
-       // NamedHasSynced delegates to the Config's Queue
-       NamedHasSynced() NamedSyncer
-
        // LastSyncResourceVersion delegates to the Reflector when there
        // is one, otherwise returns the empty string
        LastSyncResourceVersion() string
 }
 
+// NamedController adds NamedHasSynced to Controller.
+type NamedController interface {
+       Controller
+       // NamedHasSynced delegates to the Config's Queue
+       NamedHasSynced() NamedSyncer
+}
+
 // New makes a new Controller from the given Config.
+//
+// The returned implementation is guaranteed to also implement NamedController.
 func New(c *Config) Controller {
+       return newController(c)
+}
+
+func newController(c *Config) *controller {
        ctlr := &controller{
                config: *c,
                clock:  &clock.RealClock{},
diff --git a/staging/src/k8s.io/client-go/tools/cache/shared_informer.go b/staging/src/k8s.io/client-go/tools/cache/shared_informer.go
index 4f0ad8c1524..539d367e844 100644
--- a/staging/src/k8s.io/client-go/tools/cache/shared_informer.go
+++ b/staging/src/k8s.io/client-go/tools/cache/shared_informer.go
@@ -502,7 +502,7 @@ type NamedSyncer interface {
 // notifications to each of the informer's clients.
 type sharedIndexInformer struct {
        indexer    Indexer
-       controller Controller
+       controller *controller
 
        // synced gets created when creating the sharedIndexInformer.
        // It gets closed when Run detects that the processor created
@@ -670,8 +670,8 @@ func (s *sharedIndexInformer) RunWithContext(ctx context.Context) {
                        WatchErrorHandlerWithContext: s.watchErrorHandler,
                }
 
-               s.controller = New(cfg)
-               s.controller.(*controller).clock = s.clock
+               s.controller = newController(cfg)
+               s.controller.clock = s.clock
                s.syncedName.Store(ptr.To(s.controller.NamedHasSynced().Name()))
                s.started = true
        }()
```

It could be even shorter when not exposing `NamedHasSynced` outside of the package, but as `Controller` and thus `HasSynced` are public, it might make sense to give consumers access to it. Perhaps they really need it for something.

pohly replied:
Added in a separate commit because I need to rebase.

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -57,11 +65,23 @@ var _ QueueWithBatch = &RealFIFO{}
 
 	items []Delta
 
+	// synced is initially an open channel. It gets closed (once!) by hasSynced_locked
```
Only the initial sync? What about cases where we replace the underlying queue, previously this would be considered unsynced until the initial population is processed but now the channel will be closed after whatever the initial sync does. This differs from the hasSynced behavior, just want to make sure we think this through before expanding the interface. Should replace also swap out the channel?

pohly replied:
> Only the initial sync?

The usual semantic of `HasSynced` is that once an entity is synced, it remains synced. Callers which use `WaitForCacheSync` proceed as soon as they observe "true" from all entities.

I agree that this is a change of behavior, but IMHO it makes it more consistent.

> What about cases where we replace the underlying queue

Is that something that is being done?

> Should replace also swap out the channel?

Maybe, but then we have the same race as with `HasSynced` flipping back to `false` from `true`: some client waiting might have moved on already.

pohly commented:
Rebased and updated:

- Controller interface gets changed again instead of working around an unlikely API break.
- `NamedHasSynced()` -> `HasSyncedChecker()`
- `NamedSyncer` -> `DoneChecker`
- [`hasSynced_locked` fix](https://github.com/kubernetes/kubernetes/pull/135395/commits/04492c8f7296a130a11a6976cf0655872112644f) also in fifo.go 
- Improved WaitFor: log "Done waiting" as soon as possible for each checker
- Added `IsDone`: simplifies non-blocking "done" check for a `DoneChecker`
- cache/synctrack: simplified (no need for a separate interface), updated AsyncTracker the same way as SingleFileTracker
- updated staging/src/k8s.io/apiserver/pkg/admission/plugin/policy/internal/generic/controller.go to match the AsyncTracker change

I have started removing the `Controller` interface and deprecated `GetController` method. I think we should do this because it simplifies the code and might even fix a bug (garbage collector did not actually wait properly for cache sync?!) - but let's do that in a follow-up PR. We just need to make sure that we also land it in 1.35 to break the API only once.

pohly commented on staging/src/k8s.io/client-go/tools/cache/shared_informer_test.go:
Something isn't quite right yet:

```
$ go test -c k8s.io/client-go/tools/cache
$ stress ./cache.test
...
1m15s: 15 runs so far, 0 failures

/tmp/go-stress-20251219T154705-2199447272
CACHE name[0] ALTERED!
...
--- FAIL: TestAddWhileActive (3.20s)
    shared_informer_test.go:1099: Synced before processing anything?
FAIL
```

I'll check, but help is welcome.

pohly replied:
I think I found it:
```
	handle := s.processor.addListener(listener)
	for _, item := range s.indexer.List() {
		// Note that we enqueue these notifications with the lock held
		// and before returning the handle. That means there is never a
		// chance for anyone to call the handle's HasSynced method in a
		// state when it would falsely return true (i.e., when the
		// shared informer is synced but it has not observed an Add
		// with isInitialList being true, nor when the thread
		// processing notifications somehow goes faster than this
		// thread adding them and the counter is temporarily zero).
		listener.add(addNotification{newObj: item, isInInitialList: true})
	}
```

`addListener` kicks of `listener.watchSynced`, which races with the for loop. If the for loop manages to add the initial list quickly enough, all is well. If not, then `watchSynced` notices that upstream has synced at a time when there are no items to deliver, so the listener is assumed to have synced.

Moving `addListener(listener)` after the for loop would fix the race, but has the problem that `add` then may block on `p.addCh <- notification` because the consuming goroutine isn't running yet.

I think the right solution is to delay only the `listener.watchSynced`, and only in this case. It shouldn't be a problem when upstream gets started when listeners are already registered, because "upstream has synced" is required to wait until the initial list is delivered.

```patch
diff --git a/staging/src/k8s.io/client-go/tools/cache/shared_informer.go b/staging/src/k8s.io/client-go/tools/cache/shared_informer.go
index af846928a11..53dd52871da 100644
--- a/staging/src/k8s.io/client-go/tools/cache/shared_informer.go
+++ b/staging/src/k8s.io/client-go/tools/cache/shared_informer.go
@@ -836,6 +836,10 @@ func (s *sharedIndexInformer) AddEventHandlerWithOptions(handler ResourceEventHa
                // thread adding them and the counter is temporarily zero).
                listener.add(addNotification{newObj: item, isInInitialList: true})
        }
+
+       // Initial list is added, now we can allow the listener to detect that "upstream has synced".
+       s.processor.wg.Start(listener.watchSynced)
+
        return handle, nil
 }
 
@@ -955,7 +959,9 @@ func (p *sharedProcessor) addListener(listener *processorListener) ResourceEvent
        p.listeners[listener] = true
 
        if p.listenersStarted {
-               p.wg.Start(listener.watchSynced)
+               // Not starting listener.watchSynced!
+               // The caller must first add the initial list, then
+               // do it.
                p.wg.Start(listener.run)
                p.wg.Start(listener.pop)
        }
```

We could introduce a `sharedProcessor.finishAddListener` for this, but I'm not sure whether it's worth it. The caller (`sharedIndexInformer`) already has to do something special, it might as well do a bit more.

I'm going to include this patch in the commit.

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/wait_test.go:
```diff
@@ -0,0 +1,246 @@
+					default:
+						c, cancel := context.WithCancelCause(ctx)
+						wg.Go(func() {
+							time.Sleep(tc.timeout)
```
Is there no way to test this without sleeping on the timeout? Just wondering for time to complete on test.

michaelasp replied:
Gotcha, just have it engrained to flag sleeps whenever I see them. Thanks!

michaelasp commented:
Looks like there's still some test errors like in `TestSyncResult`

pohly commented:
/test pull-kubernetes-unit pull-kubernetes-integration pull-kubernetes-integration-race

Just double-checking. With `informerFactory.WaitForCacheSync` also waiting for event handler sync the previous test failures are gone.

pohly commented:
> My proposal is still to change informerFactory.WaitForCacheSync such that it does what people expect (i.e. also waits for event handler sync).

After having fixed some tests such that synctest is used to avoid other races, I am less sure how important this really is. I've taken out the commit which made this change and instead included all of my pending test fix PRs - let's see how far I can get with avoiding test flakes by fixing tests.

pohly commented:
/test pull-kubernetes-unit-canary pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

Some of the flakes for which I have PRs pending are back. But if that's all that fails, then the situation might be manageable.

pohly commented:
/test pull-kubernetes-unit

With `-parallel 0`.

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

It passed, just retesting to verify.

pohly commented:
Hmmm...

/test pull-kubernetes-unit

```
k8s.io/kubernetes/pkg/controller: job expand_less 	37s
{Failed  === RUN   TestFinalizerCleanup
    job_controller_test.go:7824: Waiting for Job object to appear in jobLister: context deadline exceeded
--- FAIL: TestFinalizerCleanup (30.02s)
```

Might be fixed by https://github.com/kubernetes/kubernetes/pull/135959. I currently don't have that in the branch for this PR.

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
The last two unit test runs had flakes, therefore I am now testing again with https://github.com/kubernetes/kubernetes/pull/135959 included.

/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
/test pull-kubernetes-unit

pohly commented:
@michaelasp: can you do another pass? I rebased, hopefully resolving conflicts correctly.

Properly fixing the List+Watch client-go race seems to have been crucial for stabilizing unit testing, I am not seeing any flakes anymore here. We might be able to merge now.

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -811,12 +839,16 @@ func newQueueFIFO(clientState Store, transform TransformFunc) Queue {
+		return f.logger, f
 	} else {
-		return NewDeltaFIFOWithOptions(DeltaFIFOOptions{
+		f := NewDeltaFIFOWithOptions(DeltaFIFOOptions{
```
We should really GA RealFIFO so we close the loop on supporting these features in DeltaFIFO. It was already agreed that we should do it in a sig-apimachinery meeting. The less duplicate code the better IMO.

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -302,9 +317,10 @@ func NewSharedIndexInformer(lw ListerWatcher, exampleObject runtime.Object, defa
 	realClock := &clock.RealClock{}
 
-	return &sharedIndexInformer{
+	s := &sharedIndexInformer{
```
Any reason we switch this, seems like we can keep it as the return.

pohly replied:
Agreed, this change isn't necessary.

I think I was experimenting with adding a logger here before settling on the current approach with `HandleDeltas` -> `handleDeltas(logger)`.

Will remove it.

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -156,11 +177,37 @@ func (f *RealFIFO) HasSynced() bool {
+// checkSynced checks whether the initial batch of items (set via Replace) has been delivered
+// and closes the synced channel as needed. It must be called after changing f.populated and/or
+// f.initialPopulationCount while the mutex is still locked.
+func (f *RealFIFO) checkSynced() {
```
This needs to have the lock when called, should we call it `checkSynced_locked()`? Just so somebody doesn't try and call this when the lock isn't held.

michaelasp reviewed (commented):
Need some more time to look through some pieces of the code but overall looks pretty good to me. I think we're close to getting this in, will look through with some dedicated time next week.

pohly commented on staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go:
```diff
@@ -132,15 +138,9 @@ func (f *sharedInformerFactory) Start(stopCh <-chan struct{}) {
-			// We need a new variable in each loop iteration,
-			// otherwise the goroutine would use the loop variable
-			// and that keeps changing.
-			informer := informer
```
Yes, loop variables always retain their value in each iteration.

michaelasp commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -311,6 +361,8 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
 	if f.initialPopulationCount > 0 {
 		f.initialPopulationCount--
+		// Must be done *after* process has completed.
+		defer f.checkSynced()
```
Yeah it doesn't have to be in this one, more just a comment on how most of this should be done after processing. At least for the other PR, we are deferring the decrement as well.

michaelasp reviewed (approved):
Looks good to me overall, minor nits at this point.

pohly commented:
Rebased because of a minor conflict with https://github.com/kubernetes/kubernetes/pull/136264.

pohly commented:
/test pull-kubernetes-kind-dra-all

"on single node supports extended resources together with ResourceClaim" - not seen elsewhere, might not run often enough yet?

pohly commented:
@michaelasp @richabanker: I've rebased on top of the informer naming changes.

Before, the name that I generated included information about ReadFIFO vs. DeltaFIFO:

```
Done waiting for="cache and event handler sync" instance="RealFIFO *v1.Pod"
```

Now it uses the `Identifier.Name()` string:
```
Done waiting for="cache and event handler sync" instance="SharedIndexInformer *v1.Pod"
```

I think that's fine. As a developer who knows the internals, "RealFIFIO" is a thing and this is relevant information, but for most developers "SharedIndexInformer" is probably better (albeit a bit unprecise).

Because there's a possibility that someone might want to name a RealFIFO without giving it an identifier, I am still adding the `Name` field, but it's only used if there's no identifier name. That's also consistent with DeltaFIFO which now has `Name` but no `Identifier` (not added there because legacy code?).

Can we try to merge this now?

aojea commented:
> How bad is it in clients? How often do they literally implement their own "HasSynced" call which now needs to be adapted vs. embedding something which has already been converted to HasSyncedChecker? Can you share some examples?

I never seen this need basically because of how controller works, I doubt a busy polling here is the main cause of slowness, think the controllers use to implement leader election and before the cache is sync it needs to perform an entire List with all that implies (get a large chunk of data from the network + deserialize + fill the cache in memory), all this happens in parallel meanwhile you are busy polling

So the critical and slow path here is the List() to populate the cache that can be very slow, or just the acquisition of the lease ... if you put those operations times in front of the busy polling you can see that there are several orders of magnitude 

> Maybe we can ease the transition with some helper.

That is what I mean, formulate hypothesis, validate somehow (how many consumers of the helper) ...that  make strong argument
